### ISSUE TITLE
Huge memory leak on mongos HA setup

### ISSUE THREAD

**Issue by jishi** [mongodb-js/mongodb-core#374]
I have been battling some weird memory leaks in our apps, and have found out that the culprit seems to be the mongodb driver which doesn't handle the Pool teardown properly when running against Mongos proxies (and maybe Replicasets as well).

This is very easily reproducible by just connecting to a set of mongos proxies (2 in my case). At app start, a heapdump shows that there are two Pool instances when everything works. If I shut down network on the machine running the driver, after 30 seconds it will start eating memory like crazy and have thousands of Pool instances in the heap. This will go on indefinitely while you have a network outage, and even when you regain network connectivity, all of these instances remains and the app uses several hundred MB of RAM. 

This was the snippet I used for testing:

```
// Simplest mongoDB connection app
const MongoClient = require('mongodb').MongoClient;

function heapDump() {
    const mem = process.memoryUsage();
    console.log('%s, Memory usage: %f', new Date().toISOString(), mem.rss);
}

async function start() {

	const client = await MongoClient.connect('mongodb://user:password@mongos1,mongos2/admin', {
		useNewUrlParser: true,
		ssl: true,
		haInterval: 10,
		bufferMaxEntries: 0,
	});
	const db = client.db(); 
}

start()
 .catch(err => console.error(err));
setInterval(heapDump, 1000);
```

The haInterval speeds things up, by default it increases with 2 Pools every 5 seconds, with this config it increases every 10 ms, making it more evident. The leak is still there. 

Not sure why it takes 30 second for the pool increase to start, but I'm guessing that is the heartbeat timeout against the servers or something. 

This was tested against mongodb@3.1.9, meaning mongodb-core@3.1.8, but I think it affect all 3.x versions (I've verified it in mongodb@3.1.4/mongo-core@3.1.3)

**Comment by mbroadst** [comment]
Can you try the suggested fix in [this comment](https://github.com/mongodb-js/mongodb-core/pull/365#issuecomment-438043421)?

**Comment by mbroadst** [comment]
@jishi thanks for testing so quickly, glad the fix seems to improve things. Is it possible the 2x factor is simply a matter of GC not running by the time you check? I’ll look into this more today.

To answer your second question: traditionally the idiom in this codebase is that objects which register event listeners should unregister themselves, but I agree in this case that “destroy” needs a stronger contract. I’ll put together a patch after lunch today for you to test. 

In other news, with the upcoming 3.2.0 release, we have completely rewritten the “topology” layer to make all of this much easier to maintain, and hopefully more performant and error free. Looking forward to having people test that out!

**Comment by mbroadst** [comment]
That’s highly likely, the reconnect code was written for the single Server case first and retrofitted to support replsets and sharded configurations. 3.2.0 aims to be released within the month, but the new SDAM implementation (topology) does not provide reconnect support out of the box (for this very reason), rather handles it through what we on drivers call “server selection”. 

If you’d like to take a sneak peak, or help us beta test, please try installing `mongodb@next` and `mongodb-core@next`, and pass `useUnifiedTopology: true` to your `MongoClient` constructor in the options. This would really help us out if you’re in a position to test! We imagine there are a number of bugs with such a large rewrite, but only minor JS oversights - the general quality wrt MongoDB should be greatly improved.

**Comment by jishi** [comment]
@mbroadst Doesn't seem like the next versions are published publicly on NPM? We have some lab-services we could easily run with experimental mongodb versions if needed.

**Comment by mbroadst** [comment]
ah sorry, I just meant the `next` branches of both projects:
`”mongodb”: “mongodb/node-mongodb-native#next”`, etc. I will most likely be publishing a `3.2.0-beta1` within a week on the `unstable` tag within a week, so that might be easier to test. Thanks for the offer of help!

**Comment by mbroadst** [comment]
fixed by #376
